components/sensor/adxl345/adxl345.go

```diff
@@ -0,0 +1,196 @@
+	if !ok {
+		return nil, errors.Errorf("can't find I2C bus '%s' for ADXL345 sensor", cfg.BusID)
+	}
+
+	var address byte
+	if cfg.UseAlternateI2CAddress {
+		address = 0x1D
+	} else {
+		address = 0x53
+	}
+
+	sensor := &adxl345{
+		bus:        bus,
+		i2cAddress: address,
+		logger:     logger,
+	}
+
+	// To check that we're able to talk to the chip, we should be able to read register 0 and get
+	// back the device ID (0xE5).
+	deviceID, err := sensor.readByte(ctx, 0)
+	if err != nil {
+		return nil, errors.Errorf("can't read from I2C address %d on bus %s of board %s: '%s'",
+			address, cfg.BusID, cfg.BoardName, err.Error())
+	}
+	if deviceID != 0xE5 {
+		return nil, errors.Errorf("unexpected I2C device instead of ADXL345 at address %d: deviceID '%d'",
+			address, deviceID)
+	}
+
+	// The chip starts out in standby mode. Set it to measurement mode so we can get data from it.
+	// To do this, we set the Power Control register (0x2D) to turn on the 8's bit.
+	err = sensor.writeByte(ctx, 0x2D, 0x08)
+	if err != nil {
+		return nil, errors.Errorf("unable to put ADXL345 into measurement mode: '%s'", err.Error())
+	}
+
+	return sensor, nil
+}
+
+func (adxl *adxl345) readByte(ctx context.Context, register byte) (byte, error) {
+	result, err := adxl.readBlock(ctx, register, 1)
+	if err != nil {
+		return 0, err
+	}
+	return result[0], err
+}
+
+func (adxl *adxl345) readBlock(ctx context.Context, register byte, length uint8) ([]byte, error) {
+	handle, err := adxl.bus.OpenHandle(adxl.i2cAddress)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		err := handle.Close()
+		if err != nil {
+			adxl.logger.Error(err)
+		}
+	}()
+
+	results, err := handle.ReadBlockData(ctx, register, length)
+	return results, err
+}
+
+func (adxl *adxl345) writeByte(ctx context.Context, register, value byte) error {
+	handle, err := adxl.bus.OpenHandle(adxl.i2cAddress)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err := handle.Close()
+		if err != nil {
+			adxl.logger.Error(err)
+		}
+	}()
+
+	return handle.WriteByteData(ctx, register, value)
+}
+
+// A helper function: takes 2 bytes and reinterprets them as a little-endian signed integer.
+func toSignedValue(data []byte) int {
+	return int(int16(binary.LittleEndian.Uint16(data)))
+}
+
+func (adxl *adxl345) Readings(ctx context.Context) (map[string]interface{}, error) {
+	adxl.mu.Lock()
+	defer adxl.mu.Unlock()
+	// The registers holding the data are 0x32 through 0x37: two bytes each for X, Y, and Z.
+	rawData, err := adxl.readBlock(ctx, 0x32, 6)
+	if err != nil {
+		return nil, err
+	}
+
+	x := toSignedValue(rawData[0:2])
+	y := toSignedValue(rawData[2:4])
+	z := toSignedValue(rawData[4:6])
+	return map[string]interface{}{"x": x, "y": y, "z": z}, nil
+}
+
+// Puts the chip into standby mode.
+func (adxl *adxl345) Close() {
+	adxl.mu.Lock()
+	defer adxl.mu.Unlock()
+	// Put the chip into standby mode by setting the Power Control register (0x2D) to 0.
+	err := adxl.writeByte(context.TODO(), 0x2D, 0x00)
```

Comment:
My new-to-Go eyes think this line looks unfinished. It's not: it originally passed in `nil` for the context, and the linter wanted me to use `context.TODO()` instead of `nil.  but it's purposely not passing any real context here.